### DISCUSSION TITLE
Using the same debug configuration in launch.json and cmake.debugConfig

### DISCUSSION THREAD

**Discussion by alexolog** [microsoft/vscode-cmake-tools#2999] [Q&A] [answered]
Currently, I have duplicate debug configurations in **launch.json**:
```
    "configurations": [
        {
            "name": "Debug CMake project",
            "type": "lldb", // https://github.com/vadimcn/vscode-lldb
            "request": "launch",
            "program": "${command:cmake.launchTargetPath}",
            "args": [
                "arg1",
                "arg2",
                "arg3",
            ]
        }
    ]
```
and in **settings.json**:
```
	"cmake.debugConfig": {
		"name": "Debug CMake project",
		"type": "lldb", // https://github.com/vadimcn/vscode-lldb
		"request": "launch",
		"program": "${command:cmake.launchTargetPath}",
		"args": [
			"arg1",
			"arg2",
			"arg3",
		]
	}
```
Both have to be maintained in tandem.

Is it possible to only have a single copy in one place and refer to it by name in the other?

If not, please consider this a feature request.

Thank you!

**Comment by bobbrow** [comment, marked as answer]
It is not expected for users to have both `launch.json` and `cmake.debugConfig`.  The CMake Tools quick launch/debug feature was designed for users who do not have a `launch.json`.  It is recommended that you use the built in VS Code UI and keyboard shortcuts (e.g. F5) to begin debugging if you have `launch.json`.

**Reply by i-ky** [reply]
If users having `launch.json` are not supposed to use quick launch/debug feature, then quick launch/debug button has to be hidden in UI.

My team is migrating to VS Code from another IDE. It is the first encounter with VS Code for many of my teammates and they try to click every button they see not knowing whether they are supposed to do it or not. If something does not work, they tend to blame me, because I did the VS Code configuration, not the extension.

**Reply by zethon** [reply]
So what is the expectation here? Should I delete my launch.json so that I can use the Debug button on the left hand side of the IDE instead of the Debug button in the status bar?

I love VSCode as an editor but the C++ debugging experience is absolutely confusing. Why do I have two different debug buttons (one on the left hand side and one on the status bar) that require two different configurations? Ok, I get the point that there are multiple C++ extensions at play, so what should I do? Disable the `MSFT - C/C++` extension and only have the `MSFT - CMake Tools` extension enabled?

**Reply by zethon** [reply]
Haha.. I just tried disabling the `MST - C/C++` extension and the Debug button in the status bar disappeared. Ok, so `MSFT - CMake Tools` does not include the C/C++ functionality.. but still WHY DO I HAVE TWO DIFFERENT DEBUG BUTTOnS THAT DO TWO DIFFERENT THINGS?

**Reply by malsyned** [reply]
@i-ky  if you're worried about new users stumbling on CMake's debug command when you want them to use the ones configured in `launch.json`, you can add the following in `settings.json`:
```json
    "cmake.options.advanced": {
        "debug": {
            "statusBarVisibility": "hidden",
            "projectStatusVisibility": "hidden"
        },
```

You can also add an identical section with `"launch"` instead of `"debug"` to remove the "CMake: Run Without Debugging" command from your project.

**Reply by Challanger524** [reply]
@malsyned you saved my day (and bar space)

```json
"cmake.options.advanced": {
    "launchTarget"  : {"statusBarVisibility": "hidden"},
    "buildPreset"   : {"statusBarVisibility": "hidden"},
    "testPreset"    : {"statusBarVisibility": "hidden"},
    "ctest"         : {"statusBarVisibility": "hidden"},
    "packagePreset" : {"statusBarVisibility": "hidden"},
    "cpack"         : {"statusBarVisibility": "hidden"},
    "workflow"      : {"statusBarVisibility": "hidden"},
    "workflowPreset": {"statusBarVisibility": "hidden"}
},
```

**Reply by malsyned** [reply]
@Challanger524 glad to hear it! I don't think it's super well-documented, but if you set both `statusBarVisibility` and `projectStatusVisibility` to `hidden`, the corresponding commands, like "CMake: Debug", are also not available via the command pallet and their keyboard shortcuts, like **Shift+F5**, are disabled.

**Reply by i-ky** [reply]
@malsyned, thank you for the suggestion! It is a nice workaround, but a workaround still. In my opinion this should be default extension's behaviour.

**Comment by oviano** [comment]
From my experience, you need both these files if you want to be able to launch the debugger from top of the CMake Project Status aswell as the Run and Debug option.

The former needs cmake.debugConfig the latter needs launch.json.

If there is a way for cmake.debugConfig to propagate its contents to launch.json, or vice-versa, I certainly haven't found it.

I'm using macOS.

**Comment by Challanger524** [comment]
1. @bobbrow, is there a way to dump/print the default `cmake.debugConfig`? Because I cannot replicate same debugger behavior and even getting application segfault when trying to debug with `msys2/clang64/sanitize=address` with `launch.json` similar to [CMake: Debug and launch](https://github.com/microsoft/vscode-cmake-tools/blob/main/docs/debug-launch.md)

2. Initially messing around with enabling `UTF-8` in **Debug Console** with no success, assuming that `internalConsole` is simply limited _to the ground_ (not only by inability to receive input).
My current solution with utf-8 output is:
```json
"cmake.debugConfig": {
	"console": "integratedTerminal",
	//"setupCommands": [{"text": "set charset UTF-8"}] // does not affect Debug Console here nor in `launch.json`
},
```
